examples/cpp/modes/goto_global/include/mode.hpp

```diff
@@ -0,0 +1,224 @@
+          const float heading_target_rad = px4_ros2::headingToGlobalPosition(
+            _vehicle_global_position->position().head(2), position_of_interest_m);
+
+          _goto_setpoint->update(_start_position_m, heading_target_rad);
+          if (positionReached(_start_position_m)) {
+            completed(px4_ros2::Result::Success);
+            return;
+          }
+        }
+        break;
+    }
+  }
+
+private:
+  static constexpr float kTriangleHeight = 20.f; // [m]
+  static constexpr float kTriangleWidth = 30.f; // [m]
+
+  enum class State
+  {
+    SettlingAtStart = 0,
+    GoingNorth,
+    GoingEast,
+    GoingSouthwest
+  } _state;
+
+  // NED earth-fixed frame. box pattern starting corner (first position the mode sees on activation)
+  Eigen::Vector3d _start_position_m;
+  bool _start_position_set{false};
+
+  // [-pi, pi] current vehicle heading from VehicleAttitude subscription
+  float _vehicle_heading_rad{0.f};
+
+  // [-pi, pi] current heading setpoint during spinning phase
+  float _spinning_heading_rad{0.f};
+
+  // used for heading initialization when dynamically updating heading setpoints
+  bool _start_heading_set{false};
+
+  std::shared_ptr<px4_ros2::GotoGlobalSetpointType> _goto_setpoint;
+  std::shared_ptr<px4_ros2::OdometryGlobalPosition> _vehicle_global_position;
+  rclcpp::Subscription<px4_msgs::msg::VehicleAttitude>::SharedPtr _vehicle_attitude_sub;
+
+  void updateVehicleHeading(const px4_msgs::msg::VehicleAttitude::UniquePtr & msg)
+  {
+    const Eigen::Quaternionf q = Eigen::Quaternionf(msg->q[0], msg->q[1], msg->q[2], msg->q[3]);
+    const Eigen::Vector3f rpy = px4_ros2::quaternionToEulerRpy(q);
+    _vehicle_heading_rad = rpy(2);
+  }
+
+  bool positionReached(const Eigen::Vector3d & target_position_m) const
+  {
+    static constexpr float kPositionErrorThreshold = 0.5f; // [m]
+    const float position_error = px4_ros2::distanceToGlobalPosition(
+      _vehicle_global_position->position(), target_position_m);
+    return position_error < kPositionErrorThreshold;
+  }
+
+  bool headingReached(float target_heading_rad) const
+  {
+    static constexpr float kHeadingErrorThreshold = 7.0_deg;
+    const float heading_error_wrapped = px4_ros2::wrapPi(target_heading_rad - _vehicle_heading_rad);
+    return fabsf(heading_error_wrapped) < kHeadingErrorThreshold;
+  }
+};
+
+class TestNode : public rclcpp::Node
+{
+public:
+  TestNode()
+  : Node(kNodeName)
+  {
+    // Enable debug output
+    auto ret =
+      rcutils_logging_set_logger_level(get_logger().get_name(), RCUTILS_LOG_SEVERITY_DEBUG);
+
+    if (ret != RCUTILS_RET_OK) {
+      RCLCPP_ERROR(get_logger(), "Error setting severity: %s", rcutils_get_error_string().str);
+      rcutils_reset_error();
+    }
+
+    _mode = std::make_unique<FlightModeTest>(*this);
+
+    if (!_mode->doRegister()) {
+      throw std::runtime_error("Registration failed");
+    }
+  }
+
+private:
+  std::unique_ptr<FlightModeTest> _mode;
+};
```

Comment:
We can start to remove this boiler-plate by adding a templated class for this, in the form of:
```
template<typename ModeT>
class NodeWithMode : public rclcpp::Node
{
     NodeWithMode(string node_name, bool enable_debug_output=false)
```
(and even better if you find a better name :)

Reply:
I'm creating a [separate PR](https://github.com/Auterion/px4-ros2-interface-lib/pull/30) for this so that there's more room for discussion, also since it's not directly related to this global/local conversion PR.